exe-common/src/sync.rs

```diff
@@ -659,3 +660,404 @@ pub fn get_chain_state_at_end_of(
         &chain_state::get_last_block_of_epoch(storage, epoch_id)?,
     )?)
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::network::Result;
+    use cardano::{
+        address::*,
+        block::{sign::*, types::*, *},
+        config::*,
+        hdwallet::*,
+        *,
+    };
+    use cbor_event::{self, de::Deserialize, de::Deserializer, se::Serializer};
+    use std::{collections::BTreeMap, str::FromStr};
+
+    // the codebase has this number hard-coded in a tonne of places
+    // so we better use the same one.
+    static SLOTS_PER_EPOCH: u16 = 21600;
+
+    struct LocalTestingPeer<'a> {
+        genesis_data: &'a GenesisData,
+        blocks: BTreeMap<HeaderHash, Block>,
+        tip: HeaderHash,
+        leader_xpub: XPub,
+        leader_xprv: XPrv,
+        // To prevent blocks that we fork and recreate from having identical hashes,
+        // we stuff a nonce into their extra_data field, or else we can never trigger a fork
+        // since the new fork will be identical to the old one (and thus have the same hashes!)
+        nonce: u64,
```

Comment:
This nonce is only included within main blocks, not boundary ones. This shouldn't be an issue as long as the forks have any main blocks at all in them for the test. We need to generate the genesis block which is why `make_boundary_block()` is a pure function.